Support multiple demo endpoints

mdube7 opened the pull request:
Added support for an id parameter after the demo to use different files. for example:
http://localhost:3000/#/demo/Maureen
http://localhost:3000/#/demo/Maryam

To add more demo records, create the new record in `pathways\public\static\demoData'

dubem7 committed:
Initial commit

parse url for id=X and use that demo record

dubem7 committed:
Fix linting issues

dehall commented on src/components/App.tsx:
```diff
@@ -134,12 +134,26 @@ const App: FC<AppProps> = ({ demo }) => {
     );
   };
 
+  function getDemoId(): number {
```
This works, but I'm wondering if there's an easier/cleaner way to do it in the router in index.js, something like a urlParam: https://reacttraining.com/react-router/web/example/url-params .

mdube7 replied:
refactored to use params

dehall commented on src/fixtures/DemoData.tsx:
```diff
@@ -0,0 +1,4 @@
+import maureenDemo from 'fixtures/MaureenMcodeDemoPatientRecords.json';
+import maureenDemo2 from 'fixtures/MaureenMcodeDemoPatientRecords2.json';
+
+export const DemoRecords = [maureenDemo, maureenDemo2];
```
Again this definitely works, but there's the downside of we have to make code changes every time we want to add a new demo file. Maybe it's not possible with the files being in the `src` folder, but could you look into doing a `fetch` for patient records under the `public/static` folder, similar to what we do for CQL libraries? We could enforce a filename pattern or just say file `maureenDemo.json` gets read when you go to `http://pathways/#/demo/maureenDemo`

mdube7 replied:
refactored to fetch records from public/static

blangley28 commented on src/fixtures/DemoData.tsx:
```diff
@@ -0,0 +1,4 @@
+import maureenDemo from 'fixtures/MaureenMcodeDemoPatientRecords.json';
+import maureenDemo2 from 'fixtures/MaureenMcodeDemoPatientRecords2.json';
+
+export const DemoRecords = [maureenDemo, maureenDemo2];
```
Up for discussion since this is the only export in the module but `DemoRecords` is a variable so I think it should be `demoRecords`. For reference the code style guide is here: https://github.com/mcode/pathways/wiki/Code-Style-Guide

mdube7 replied:
file was removed

blangley28 commented on src/components/App.tsx:
```diff
@@ -134,12 +134,26 @@ const App: FC<AppProps> = ({ demo }) => {
+    // parse the url for the id of the patient to use
+    const url = window.document.location.href;
+    const params = url.split('?');
+    const urlParams = new URLSearchParams(params[1]);
```
You don't need to use split here. The following should work
`const urlParams = new URLSearchParams(window.location.search);`

dubem7 committed:
Use useParams in index.js for the id

Load the file that was specified in the url

dubem7 committed:
Adding support for multiple endpoints

Using fetch to get the demo data from public/static

Commit:
Updated PatientProvider.

dubem7 committed:
Force Refresh on url change, some cleanup

dubem7 committed:
Remove hard-coded URL and Code Cleanup

dubem7 committed:
Fix linting errors

dubem7 committed:
Remove commented import and console.log

dubem7 committed:
Remove key from default demo url

blangley28 commented on src/index.js:
```diff
@@ -11,8 +11,17 @@ const rootElement = document.getElementById('root');
     <Switch>
-      <Route exact path="/" render={() => <App demo={false} />}></Route>
-      <Route path="/demo" render={() => <App demo={true} />}></Route>
+      <Route exact path="/" render={() => <App demo={false} demoData={''} />}></Route>
```
`demoData` should probably just be made optional instead of passing in an empty string

blangley28 commented on src/components/App.tsx:
```diff
@@ -18,15 +18,16 @@ import ThemeProvider from './ThemeProvider';
 import { Patient, DomainResource, Practitioner } from 'fhir-objects';
 interface AppProps {
   demo: boolean;
+  demoId: string;
```
Make this optional 
```suggestion
  demoId?: string;
```

blangley28 commented on src/components/PatientProvider.tsx:
```diff
@@ -1,27 +1,29 @@
-  const [currentPatient, setCurrentPatient] = useState<Patient | null>(patient || null);
+interface PatientContextInterface {
+  patient: Patient | null;
+  setPatient: Function;
```
Maybe I am missing something but it seems like `setPatient` is never used in this context. The only time it is called is in the `App` component but that is using the `setPatient` state method created in `App` not from the context. Is there a reason this changed?

danlee1025 replied:
This was changed to follow the same pattern as `PatientRecordsProvider`. The `patient` state was moved into the `App` component so that it would know about changes to the patient. It seems like no other component currently needs the `setPatient` function so we could remove it but in the future if other components need to change the patient we would need to add this function back.

dubem7 committed:
Update Patient Record after rebase

dehall commented on public/config.js:
```diff
@@ -2,5 +2,6 @@ CONFIG = {
   pathwaysService: 'http://pathways.mitre.org:3002/pathways/',
-  demoPathwaysService: './static/pathways/'
+  demoPathwaysService: './static/pathways/',
+  demoPatientUrl: './static/demoData/'
```
Can we call this `demoPatients` to be more specific?

mdube7 replied:
Updated name

dehall commented on src/components/App.tsx:
```diff
@@ -18,17 +18,17 @@ import ThemeProvider from './ThemeProvider';
-
 interface AppProps {
   demo: boolean;
+  demoId: string;
```
We could theoretically condense these into one field... `demoId` being an optional string, if it's null then use the FHIR client. Not super high priority

mdube7 replied:
removed demo

dehall commented on src/components/App.tsx:
```diff
@@ -65,13 +65,21 @@ const App: FC<AppProps> = ({ demo }) => {
+        .then(data => data.json())
+        .then(result => {
+          const resultPatient = result.find((r: DomainResource) => r.resourceType === 'Patient');
+          setPatientRecords(result);
```
For future consideration only, not here, I wonder if it makes sense to combine the PatientRecordsProvider and PatientProvider

danlee1025 replied:
Agreed, we didn't always have Context for both and when I decided to create the `PatientRecords` one I didn't think to combine them.

dehall reviewed (commented):
Overall looks good. I feel like there are probably ways it could be cleaned up a little further, but I don't see anything specifically.
One thought: we should probably pick a naming convention for demo patient files. Originally I thought we would just use numbers but maybe someone has a better idea. Since we have `/demo/` in the URL I don't think we need `Demo` in the filename too though.

Also could you replace MaureenDemo2 with this patient record? It's more different and is actually FHIR R4.
[maryam.json.txt](https://github.com/mcode/pathways/files/4429247/maryam.json.txt)

dubem7 committed:
Address PR issues

Change naming convention of demo patients
remove the demo prop
change demoPatientUrl to demoPatients

blangley28 commented on src/components/App.tsx:
```diff
@@ -84,7 +91,7 @@ const App: FC<AppProps> = ({ demo }) => {
 
   const service = useGetPathwaysService(
-    config.get(demo ? 'demoPathwaysService' : 'pathwaysService')
+    config.get(demoId === null ? 'demoPathwaysService' : 'pathwaysService')
```
This is inconsistent with the prior behavior - `demo === true` should transform to `demoId !== null`
```suggestion
    config.get(demoId !== null ? 'demoPathwaysService' : 'pathwaysService')
```

mdube7 replied:
updated to !==

blangley28 commented on src/components/ExpandedNode/ExpandedNode.tsx:
```diff
@@ -44,7 +44,7 @@ const ExpandedNode: FC<ExpandedNodeProps> = ({
   const client = useFHIRClient();
   const note = useNote();
-  const patient = usePatient();
+  const patient = usePatient().patient as fhir.Patient;
```
The type of `usePatient().patient` is `fhir.Patient | null`. I think all this `as fhir.Patient` is doing is covering a null check. In my personal opinion we should make the null check explicit because it is easier to read but would like to hear from others too.

danlee1025 replied:
I mean by the time we reach these components, `patient` will have already been set since the app will produce a loading screen instead of rendering the `children` components when `patient` is `null`.

blangley28 commented on src/index.js:
```diff
@@ -11,8 +11,12 @@ const rootElement = document.getElementById('root');
     <Switch>
-      <Route exact path="/" render={() => <App demo={false} />}></Route>
-      <Route path="/demo" render={() => <App demo={true} />}></Route>
+      <Route exact path="/" render={() => <App demoData={null} />}></Route>
```
Looks like `demoData` was removed from the props interface so it should be deleted here
```suggestion
      <Route exact path="/" render={() => <App  />}></Route>
```

mdube7 replied:
updated to demoId={null}

blangley28 replied:
`demoId` is optional so it can be removed entirely and just be `<App />`

blangley28 commented:
The app doesn't seem to launch correctly from the smart launcher
![Screen Shot 2020-04-06 at 10 02 19 AM](https://user-images.githubusercontent.com/53485517/78566774-b7090e80-77ed-11ea-984a-0ed22a55a3c0.png)

dubem7 committed:
Address issues in PR

mdube7 commented:
> The app doesn't seem to launch correctly from the smart launcher

The smart launcher is working again now

dehall reviewed (approved):
Looks good, the only thing that seems a little weird is that if you start at http://localhost:3000/#/ and then go to http://localhost:3000/#/demo without refreshing, it uses the pathways from the service URL instead of from the demo location. I think that's just an artifact of how we're loading the pathway service URL and we can revisit later if that ever becomes an issue
